src/workspace.coffee

```diff
@@ -791,36 +800,76 @@ class Workspace extends Model
   # * `regex` {RegExp} to search with.
   # * `options` (optional) {Object} (default: {})
   #   * `paths` An {Array} of glob patterns to search within
+  #   * `onPathsSearched` (optional) {Function}
   # * `iterator` {Function} callback on each file found
   #
-  # Returns a `Promise`.
+  # Returns a *thenable* object with a `cancel()` method that will cancel all
+  # of the underlying searches that were started as part of this scan.
   scan: (regex, options={}, iterator) ->
     if _.isFunction(options)
       iterator = options
       options = {}
 
-    deferred = Q.defer()
-
     searchOptions =
+      regexSource: regex.source
       ignoreCase: regex.ignoreCase
-      inclusions: options.paths
+      inclusions: options.paths or []
       includeHidden: true
       excludeVcsIgnores: atom.config.get('core.excludeVcsIgnoredPaths')
       exclusions: atom.config.get('core.ignoredNames')
       follow: atom.config.get('core.followSymlinks')
 
-    task = Task.once require.resolve('./scan-handler'), atom.project.getPaths(), regex.source, searchOptions, ->
-      deferred.resolve()
-
-    task.on 'scan:result-found', (result) ->
-      iterator(result) unless atom.project.isPathModified(result.filePath)
+    # Find a searcher for every Directory in the project.
+    searchersAndDirectories = []
+    for directory in atom.project.getDirectories()
+      searcher = null
+      for directorySearcher in @directorySearchers
+        if directorySearcher.canSearchDirectory(directory)
+          searcher = directorySearcher
+          break
+      if searcher
+        searchersAndDirectories.push({searcher, directory})
+      else
+        throw Error("Could not find directory searcher for #{directory.getPath()}")
```

Comment:
This is similar to what you did for the directory-provider right, where this exception is just here to signal an internal error? I was thinking it might be more clear and direct to explicitly choose the `DefaultDirectorySearcher` here, rather than putting it into the `@directorySearchers` array. Then we wouldn't need this `throw` here at all.